src/application/service/ExchangeAuthCode.ts

```diff
@@ -0,0 +1,53 @@
+import { MagicLinkRepository } from '../../domain/model/magiclink/MagicLinkRepository';
+import { GenerateAuthToken } from './GenerateAuthToken';
+import { GetExistingOrCreateNewUser } from './GetExistingOrCreateNewUser';
+
+export class ExchangeAuthCode {
+  constructor(
+    private magicLinkRepository: MagicLinkRepository,
+    private generateAuthToken: GenerateAuthToken,
+    private getExistingOrCreateNewUser: GetExistingOrCreateNewUser
+  ) {}
+
+  public async execute(email: string, authCode: string) {
+    const magicLink = await this.magicLinkRepository.findByCode(authCode);
+
+    if (!magicLink || magicLink.email.value !== email) {
+      throw new AuthorisationFailedError(
+        AuthorisationFailureReason.AUTH_CODE_OR_EMAIL_NOT_FOUND
+      );
+    }
+
+    if (magicLink.isExpired()) {
+      throw new AuthorisationFailedError(
+        AuthorisationFailureReason.AUTH_CODE_EXPIRED
+      );
+    }
+
+    if (!magicLink.active) {
+      throw new AuthorisationFailedError(
+        AuthorisationFailureReason.AUTH_CODE_ALREADY_USED
+      );
+    }
+
+    const user = await this.getExistingOrCreateNewUser.execute(email);
```

Comment:
Uhmm... I'm not sure we should do all of this in this same call.

In my head (and I think in the discussions we also had in slack) the flow is more like:

* User fills a form with the email
* User receives an email
* User clicks link in the email and goes to a page where we show the registration form (name, dob, address, etc)
* User clicks submit. It's then when we create the user, and exchange the link for a JWT token

If we don't do it like that, we are gonna make more complicated the logic later, as we need to take also in consideration when an user exists, but has not really filled the information

Reply:
Not sure  how what you are proposing is any different in terms of the need to check if a user exists and if not create it here. No matter if we allow null profile/address or not this logic will still be needed. 

Furthermore I believe posting profile information on an auth/login endpoint sounds wrong. I'd much rather we decouple those endpoints even if that means checking in the app if the user has profile information after login.